PowerEditor/src/Notepad_plus.cpp

```diff
@@ -3380,15 +3360,15 @@ bool isUrl(TCHAR * text, int textLen, int start, int* segmentLen)
 			URL_COMPONENTS url;
 			memset (& url, 0, sizeof(url));
 			url.dwStructSize = sizeof(url);
-			bool r  = InternetCrackUrl(& text [start], len, 0, & url) && isUrlSchemeSupported(url.nScheme, & text [start]);
+			bool r  = InternetCrackUrl(& text [start], len, 0, & url);
 			if (r)
 			{
 				while (removeUnwantedTrailingCharFromUrl (& text [start], & len));
 				*segmentLen = len;
 				return true;
 			}
 		}
-		len = 1;
+		else len = 1;

```

Comment:
Hmm... this line has changed the semantic of this function.
In your PR, if `r` is FALSE, then `len` will keep its value, and then finish the function with the evaluated value by returning false.
Is it your intention? 
If it is, could you provide brief explanation, and make `else len = 1;` as
```cpp
else
{
    len = 1;
}
```
so the intention will not be unnoticed.

Reply:
It is intended — it is, in fact, the entire point of the change.

Issue #13916 happened because URLs can contain other URLs, and the previous process for finding URLs could wind up scanning the same text within a very long line very many times if something that "looked like" a URL but wasn’t, contained many other strings that also “looked like” URLs but were not.

This PR does two things.

First, it tightens the requirements for “looks like a URL” by checking the scheme earlier in the process. That is necessary to keep the next step from skipping valid URLs in reasonable contexts.

Second, once the beginning of a potential URL passes the tighter initial scanning and the end of the URL is found, we “commit” to that portion of the line. If the potential URL fails InternetCrackUrl validation, we restart scanning from the end of of the string that looked like a URL but wasn’t, rather than from just after the scheme.

In rare cases, this can cause a valid URL that “looks like” part of an invalid URL to be missed; for example, in the text ```ftp://this:4k,ftp://stop``` the existing logic underlines ```ftp://stop```; but with this PR nothing is underlined. I believe that is the unavoidable cost of preventing the “backtracking” loop observed in #13916.

In attempting to explain this, I have found a subtle error. The fall-through to the `while` loop that checks `isUrlSchemeStartChar` is usually unnecessary; that loop avoids restarting the search for a scheme in the middle of a longer string of alphabetic characters. The only time the end of a (potential) URL falls just before an alphabetic character is when the URL was ended by a delimited query string. In that situation — e.g., `ftp://this:4k?that="12"ftp://stop` — the second ftp *should* be recognized as the start of a URL (since it was not part of the invalid URL already scanned), but my change causes it to be skipped.

I will re-examine and retest; I think the correction, which will also be clearer, will be to add an `else` clause to the `if (r)` which will return false and the correct segment length — thus avoiding the fall-through — and remove the `else` I added before `len = 1;`. I’ll add a commit to this branch when I’m sure.